crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{sync::Arc, time::Duration};
+
+use futures::stream::FuturesOrdered;
+use sui_metrics::spawn_monitored_task;
+use sui_types::{
+    base_types::{AuthorityName, ExecutionDigests, TransactionDigest},
+    committee::StakeUnit,
+    error::{SuiError, SuiResult},
+    messages::{TransactionEffects, VerifiedCertificate},
+    messages_checkpoint::{CheckpointSequenceNumber, VerifiedCheckpoint},
+};
+use tokio::{
+    sync::broadcast::{self, error::RecvError},
+    task::JoinHandle,
+};
+use tokio_stream::StreamExt;
+use tracing::{error, info, warn};
+use typed_store::rocks::TypedStoreError;
+
+use crate::{
+    authority::{AuthorityState, EffectsNotifyRead},
+    checkpoints::{CheckpointMetrics, CheckpointStore},
+};
+
+const TASKS_PER_CORE: usize = 1;
+
+pub struct CheckpointExecutor {
+    mailbox: broadcast::Receiver<VerifiedCheckpoint>,
+    checkpoint_store: Arc<CheckpointStore>,
+    authority_state: Arc<AuthorityState>,
+    highest_executed_checkpoint: Option<CheckpointSequenceNumber>,
+    metrics: Arc<CheckpointMetrics>,
+}
+
+impl CheckpointExecutor {
+    pub fn new(
+        state_sync_handle: &sui_network::state_sync::Handle,
+        checkpoint_store: Arc<CheckpointStore>,
+        authority_state: Arc<AuthorityState>,
+        metrics: Arc<CheckpointMetrics>,
+    ) -> Result<Self, TypedStoreError> {
+        let mailbox = state_sync_handle.subscribe_to_synced_checkpoints();
+        let highest_executed_checkpoint =
+            checkpoint_store.get_highest_executed_checkpoint_seq_number()?;
+
+        Ok(Self {
+            mailbox,
+            checkpoint_store,
+            highest_executed_checkpoint,
+            authority_state,
+            metrics,
+        })
+    }
+
+    pub async fn run(mut self) {
+        let mut finished: FuturesOrdered<JoinHandle<(u64, VerifiedCheckpoint)>> =
+            FuturesOrdered::new();
+        // let limit = Arc::new(Semaphore::new(num_cpus::get()));
+        let task_limit = TASKS_PER_CORE * num_cpus::get();
+
+        loop {
+            tokio::select! {
+                // Check for completed workers and ratchet the highest_checkpoint_executed
+                // watermark accordingly. Note that given that checkpoints are guaranteed to
+                // be processed (added to FuturesOrdered) in seq_number order, using FuturesOrdered
+                // guarantees that we will also ratchet the watermarks in order.
+                Some(Ok((seq_num, checkpoint))) = finished.next() => {
+                    self.highest_executed_checkpoint = Some(seq_num as CheckpointSequenceNumber);
```

Comment:
@andll re: shutdown safety -- `CheckpointExecutor` sets a new watermark, `highest_executed_checkpoint`, which is persisted in the checkpoint store once we receive notification that txn effects have been persisted for a checkpoint. In the case of a shutdown, or in general, we will always start execution from `highest_executed_checkpoint + 1`